Improvement/update for cool multiaccount

jsf9k opened the pull request:
## 🗣 Description

In this pull request I update this repo to work with our current implementation of the COOL multi-account environment.

## 💭 Motivation and Context

This code is to be deployed to the COOL.  Therefore it must work with our COOL multi-account environment.

## 🧪 Testing

I deployed these changes to production, and I verified that all the pre-commit hooks passed.

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

jsf9k committed:
Modify backend and providers for COOL multi-account

jsf9k committed:
Add *.tfvars files to .gitignore

jsf9k committed:
Rename repo to agree with the way we have been naming things for the COOL

jsf9k committed:
Add policy for provisioning OpenVPN

jsf9k committed:
Add new variables for the OpenVPN provisioning policy

jsf9k committed:
Remove usage section, since this is a deployment and not a module

jsf9k committed:
Change capitalization in default value for variable

jsf9k committed:
Rename some providers

jsf9k committed:
Determine the shared services account ID via the default provider's caller identity

We used to determine this from a role ARN, but we can no longer do
that since we switched to using profiles instead.

jsf9k committed:
Rename a few providers

jsf9k committed:
Merge remote-tracking branch 'origin/develop' into improvement/update-for-cool-multiaccount

dav3r commented on provisionopenvpn_policy.tf:
```diff
@@ -0,0 +1,23 @@
+# ------------------------------------------------------------------------------
+
+data "aws_iam_policy_document" "provisionopenvpn_policy_doc" {
+  # This policy needs to be more restricted, obviously
```
🎗Reminder to create an issue to go back and restrict this policy (if this PR is merged and this policy is still wide open).

dav3r reviewed (approved):
So far, so good! 👍 
I left a familiar-looking reminder in my one comment...

felddy reviewed (approved):
Looks good sir.  I'm getting the tingles. 🎇

jsf9k committed:
Specify the go version

This is required to get around the fact that terraform-docs does not
install with the default version of go that is present.

jsf9k committed:
Go ahead and start using v2 of the setup-go action, even though it's only in beta right now

jsf9k committed:
Merge pull request #14 from cisagov/bugfix/fix-broken-terraform-docs-install

Fix broken terraform-docs install

jsf9k committed:
Merge remote-tracking branch 'skeleton-tf-module/develop' into improvement/update-for-cool-multiaccount

jsf9k committed:
Correct the default cert bucket name for the COOL

jsf9k committed:
Add the COOL Images account ID as the AMI owner

jsf9k committed:
Edit the TF foo so that the subdomain comes out correctly

jsf9k committed:
Use the FQDN for the hostname, and drop the subdomain parameter

jsf9k committed:
Add remote state for cool-sharedservices-{freeipa,networking}

This allow us to get rid of some input variables.

jsf9k committed:
Add routes to direct VPN client traffic to the OpenVPN instance

@felddy noticed that this was needed and created
cisagov/cool-sharedservices-networking#11.

dav3r commented on openvpn.tf:
```diff
@@ -30,16 +30,16 @@ module "openvpn" {
-  private_reverse_zone_id = var.private_reverse_zone_id
-  private_zone_id         = var.private_zone_id
+  private_networks        = [data.terraform_remote_state.networking.outputs.vpc.cidr_block]
+  private_reverse_zone_id = data.terraform_remote_state.networking.outputs.public_subnet_private_reverse_zones["10.128.9.0/24"].id
```
Can `"10.128.9.0/24"` be replaced with a variable here and below?

jsf9k replied:
See commit 5996486.

dav3r replied:
Gravy! 👍

jsf9k committed:
Calculate the public subnet CIDR instead of hardcoding it

jsf9k committed:
Add back in the private networks variable

The private network is the COOL as a whole, which is larger than the
SharedServices VPC CIDR block.  (Duh.)

jsf9k committed:
Use the numerically smallest subnet

The output of keys() is sorted lexicographically, so we have to do
some work to determine the correct subnet.

jsf9k committed:
Specify the string type for variables

Previously string was the default type, but this is no longer the
case:
https://www.terraform.io/docs/configuration/variables.html#type-constraints

I ran into a case today where this mattered.  I used
set([var.string_value]) in a for_each, and Terraform rejected the
syntax since it couldn't verify that the output would be a set of
strings.  Adding the string type declaration to the variable fixed the
issue.

jsf9k committed:
Revert "Add routes to direct VPN client traffic to the OpenVPN instance"

This reverts commit f7ffb60c4b02f3cdb3b2ac77ad6c2c352e602e10.

The routing approach did not pan out, so we are just going to use NAT
- at least for now.

felddy committed:
Add variables for new client DNS variables.

felddy committed:
Declare new variables as strings.

felddy committed:
Merge pull request #4 from cisagov/improvement/push_client_dns

Add variables for new client DNS variables.
